ERCS/erc-7836.md

```diff
@@ -0,0 +1,188 @@
+---
+eip: 7836
+title: Wallet Call Preparation API
+description: Adds JSON-RPC methods for requesting prepared EIP-5792 calls.
+author: Lukas Rosario (@lukasrosario), Conner Swenberg (@ilikesymmetry), Adam Hodges (@ajhodges), Paaras Bhandari (@paarasbhandari)
+discussions-to: https://ethereum-magicians.org/t/new-erc-wallet-call-preparation-api/22456
+status: Draft
+type: Standards Track
+category: ERC
+created: 2024-12-06
+requires: 1193, 5792
+---
+
+## Abstract
+
+This proposal defines complementary JSON-RPC methods to [EIP-5792](./eip-5792.md)'s `wallet_sendCalls`. While `wallet_sendCalls` is used for an app to submit `calls` to be signed and submitted in a wallet's interface, the methods in this proposal are for an application to request prepared calls (where "prepared" depends on the wallet implementation) to be signed and submitted by the application itself.
+
+## Motivation
+
+With more recent developments in account abstraction and session keys, there is an increasing need for applications to be able to sign and submit operations without switching to a wallet interface (e.g. browser extension). This can be tricky because different account implementations might have different call data encoding, signature formats, etc. To remedy this, we need a way for apps to know how to submit an operation for any account implementation.
+
+## Specification
+
+Two new JSON-RPC methods are defined. If the user is interacting with an application on the browser, they can be used by sending the requests to the wallet's [EIP-1193](./eip-1193.md) provider. However, this proposal also describes how an application might interact with a "wallet server" for background processes.
+
+### `wallet_prepareCalls`
+
+Accepts an EIP-5792 `wallet_sendCalls` request, and returns the prepared calls according to the account's implementation.
+
+#### Parameters
+
+```typescript!
+type PrepareCallsParams = [{
+  version: string
+  from: `0x${string}`
+  chainId: `0x${string}`
+  calls: {
+    to: `0x${string}`
+    data: `0x${string}`
+    value: `0x${string}`
+  }[];
+  capabilities: Record<string, any>
+}]
+```
+
+##### Example Parameters
+
+```typescript!
+wallet_prepareCalls([{
+  version: '1.0',
+  from: '0x...',
+  chainId: '0x...',
+  calls: [{
+    to: '0x...'
+    data: '0x...'
+    value: '0x...'
+  }],
+  capabilities: {
+    paymasterService: {
+      url: 'https://...'
+    }
+  }
+}])
+```
+
+#### Return Value
+
+```typescript!
+type PrepareCallsReturnValue = {
+  type: string
+  chainId: `0x${string}`
+  signatureRequest: {
+    hash: `0x${string}`
+  }
+  capabilities: Record<string, any>
+}
+##### Example Return Value
+
+```typescript!
+{
+  type: 'user-operation-v06',
+  userOp: {
+    sender: '0x...',
+    ...
+  },
+  chainId: '0x01',
+  signatureRequest: {
+    hash: '0x...'
+  }
```

Comment:
@adamegyed and I were chatting about this field, it'd be nice if this was something like:
```ts
signatureRequest: {
  type: "personal_sign" | "eth_signTypedData_v4" | ...,
  data: any, // this would match the params for the above type
}
```

As an example: for EP 0.8 where the signature of a UO is a 712 signature this could allow a client to compare the signature request to the returned `data` field as well to make it less of a blind sign.

It might depend on the client, so maybe webauthn or other signature types are required

Reply:
Why is this neccessary? You can detect the type via the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) prefix (`0x19 <1 byte version> <version specific data> <data to sign>`).